keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -70,35 +72,33 @@ class ResourceActuator(
       }
 
       try {
-        val (desired, current) = plugin.resolve(resource)
+        val (resolvedResource, desired, current) = plugin.resolve(resource)
         val diff = DefaultResourceDiff(desired, current)
         if (diff.hasChanges()) {
           diffFingerprintRepository.store(id, diff)
         }
 
+        /**
+         * [VersionedArtifactContainer] is a special [ResourceSpec] sub-type. When the resource under evaluation
+         * has a spec of this type, it means it carries information about a delivery artifact and version (typically
+         * compute resources).
+         */
+        val versionedArtifact = if (resolvedResource.spec is VersionedArtifactContainer) {
+          resolvedResource.spec as VersionedArtifactContainer
+        } else {
+          null
+        }
+
         val response = vetoEnforcer.canCheck(resource)
         if (!response.allowed) {
           /**
-           * [VersionedArtifactContainer] is a special [resource] sub-type. When a veto response sets
-           * [VetoResponse.vetoArtifact] and the resource under evaluation is of type
-           * [VersionedArtifactContainer], blacklist the desired artifact version from the environment
-           * containing [resource]. This ensures that the environment will be fully restored to
-           * a prior good-state.
+           * When a veto response sets [VetoResponse.vetoArtifact] and the resource has artifact information, blacklist
+           * the desired artifact version from the environment containing [resource]. This ensures that the environment
+           * will be fully restored to a prior good-state.
            */
-          if (response.vetoArtifact && resource.spec is VersionedArtifactContainer) {
+          if (response.vetoArtifact && versionedArtifact != null) {
             try {
-              val (version, artifact) = when (desired) {
-                is Map<*, *> -> {
-                  if (desired.size > 0) {
-                    val versioned = (desired as Map<String, VersionedArtifactContainer>).values.first()
-                    Pair(versioned.artifactVersion, versioned.deliveryArtifact)
-                  } else {
-                    Pair(null, null)
-                  }
-                }
-                is VersionedArtifactContainer -> Pair(desired.artifactVersion, desired.deliveryArtifact)
-                else -> Pair(null, null)
-              }
+              val (version, artifact) = versionedArtifact.artifactVersion to versionedArtifact.deliveryArtifact
```

Comment:
With the resolved resource available above, we no longer need to look into the concrete type to find the artifact info, so I was able to simplify this block.